File: 3741-minimum-distance-between-three-equal-elements-ii/3741-minimum-distance-between-three-equal-elements-ii.cpp
class Solution {
public:
    int minimumDistance(vector<int>& nums) {
        unordered_map<int, vector<int>> mp;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            mp[nums[i]].push_back(i);
        }

        int mini = INT_MAX;

        for (auto& it : mp) {
            vector<int>& temp = it.second;
            int m = temp.size();

            if (m >= 3) {
                for (int i = 0; i < m - 2; i++) {
                    int a = temp[i];
                    int b = temp[i + 1];
                    int c = temp[i + 2];

                    int diff = 2 * (max({a, b, c}) - min({a, b, c}));
                    mini = min(mini, diff);
                }
            }
        }

        return mini == INT_MAX ? -1 : mini;
    }
};

File: 3741-minimum-distance-between-three-equal-elements-ii/3741-minimum-distance-between-three-equal-elements-ii.java
class Solution {
    public int minimumDistance(int[] nums) {
        int n = nums.length;
        int ans = Integer.MAX_VALUE;
        int prev1[] = new int[n+1];
        int prev2[] = new int[n+1];
        for(int i=0;i<n+1;i++){
            prev1[i] = prev2[i] = -1;
        }
        
        for(int i=0;i<n;i++){
            int value = nums[i];
            if(prev2[value] != -1){
                ans = Math.min(ans, (i-prev2[value]));
            }
            prev2[value] = prev1[value];
            prev1[value] = i;
        }
        return (ans==Integer.MAX_VALUE)? -1: 2*ans;
    }
}